trio/_file_io.py

```diff
@@ -48,59 +74,199 @@
+        def raw(self) -> io.RawIOBase: ...
+
+    class _HasLineBuffering(Protocol):
+        @property
+        def line_buffering(self) -> bool: ...
+
+    class _HasCloseFD(Protocol):
+        @property
+        def closefd(self) -> bool: ...
+
+    class _HasName(Protocol):
+        @property
+        def name(self) -> str: ...
+
+    class _HasMode(Protocol):
+        @property
+        def mode(self) -> str: ...
+
+    class _CanGetValue(Protocol[AnyStr_co]):
+        def getvalue(self) -> AnyStr_co: ...
+
+    class _CanGetBuffer(Protocol):
+        def getbuffer(self) -> memoryview: ...
+
+    class _CanFlush(Protocol):
+        def flush(self) -> None: ...
+
+    class _CanRead(Protocol[AnyStr_co]):
+        def read(self, __size: int | None = ...) -> AnyStr_co: ...
+
+    class _CanRead1(Protocol):
+        def read1(self, __size: int | None = ...) -> bytes: ...
+
+    class _CanReadAll(Protocol[AnyStr_co]):
+        def readall(self) -> AnyStr_co: ...
+
+    class _CanReadInto(Protocol):
+        def readinto(self, __buf: Buffer) -> int | None: ...
+
+    class _CanReadInto1(Protocol):
+        def readinto1(self, __buffer: Buffer) -> int: ...
+
+    class _CanReadLine(Protocol[AnyStr_co]):
+        def readline(self, __size: int = ...) -> AnyStr_co: ...
+
+    class _CanReadLines(Protocol[AnyStr]):
+        def readlines(self, __hint: int = ...) -> list[AnyStr]: ...
+
+    class _CanSeek(Protocol):
+        def seek(self, __target: int, __whence: int = 0) -> int: ...
+
+    class _CanTell(Protocol):
+        def tell(self) -> int: ...
+
+    class _CanTruncate(Protocol):
+        def truncate(self, __size: int | None = ...) -> int: ...
+
+    class _CanWrite(Protocol[AnyStr_contra]):
+        def write(self, __data: AnyStr_contra) -> int: ...
+
+    class _CanWriteLines(Protocol[T_contra]):
+        """The lines parameter varies for bytes/str, so use a typevar to make the async match."""
+        def writelines(self, __lines: Iterable[T_contra]) -> None: ...
+
+    class _CanPeek(Protocol[AnyStr_co]):
+        def peek(self, __size: int = 0) -> AnyStr_co: ...
+
+    class _CanDetach(Protocol[T_co]):
+        """The T typevar will be the unbuffered/binary file this file wraps."""
+        def detach(self) -> T_co: ...
+
+    class _CanClose(Protocol):
+        def close(self) -> None: ...
+
+
+class AsyncIOWrapper(AsyncResource, Generic[FileT_co]):
     """A generic :class:`~io.IOBase` wrapper that implements the :term:`asynchronous
     file object` interface. Wrapped methods that could block are executed in
     :meth:`trio.to_thread.run_sync`.
 
-    All properties and methods defined in in :mod:`~io` are exposed by this
+    All properties and methods defined in :mod:`~io` are exposed by this
     wrapper, if they exist in the wrapped file object.
-
     """
 
-    def __init__(self, file):
+    # FileT needs to be covariant for the protocol trick to work - the real IO types are a subtype
+    # of the protocols.
```

Comment:
I think this belongs above the `class` definition line (where you first subclass from `Generic[FileT_co]`.